math/src/main/java/smile/math/matrix/SparseMatrix.java

```diff
@@ -175,11 +180,204 @@ public int size() {
     public double[] values() {
         return x;
     }
-    
+
+    /**
+     * Returns the row index
+     */
+    public int[] getRowIndex() {
+        return rowIndex;
+    }
+
+    /**
+     * Returns the column index
+     */
+    @SuppressWarnings("WeakerAccess")
+    public int[] getColIndex() {
+        return colIndex;
+    }
+
+    /**
+     * Calls a lambda with each non-zero value. This will be a bit faster than using a stream
+     * because we can avoid boxing up the coordinates of the element being processed, but it
+     * will be considerably less general.
+     * <p>
+     * Note that the action could be called on values that are either effectively or actually
+     * zero. The only guarantee is that no values that are known to be zero based on the
+     * structure of the matrix will be processed.
+     *
+     * @param action Action to perform on each non-zero, typically this is a lambda.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public void foreachNonzero(MatrixElementConsumer action) {
+        for (int j = 0; j < ncols; j++) {
+            for (int k = colIndex[j]; k < colIndex[j + 1]; k++) {
+                int i = rowIndex[k];
+                action.eval(i, j, x[k]);
+            }
+        }
+    }
+
+
+    /**
+     * Calls a lambda with each non-zero value. This will be a bit faster than using a stream
+     * because we can avoid boxing up the coordinates of the element being processed, but it
+     * will be considerably less general.
+     * <p>
+     * Note that the action could be called on values that are either effectively or actually
+     * zero. The only guarantee is that no values that are known to be zero based on the
+     * structure of the matrix will be processed.
+     *
+     * @param startColumn The first column to scan.
+     * @param endColumn   One past the last column to scan.
+     * @param action      Action to perform on each non-zero, typically this is a lambda.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public void foreachNonzero(int startColumn, int endColumn, MatrixElementConsumer action) {
+        if (startColumn < 0 || startColumn >= ncols) {
+            throw new IllegalArgumentException("Start column must be in range [0,ncols)");
+        }
+        if (endColumn < 0 || endColumn > ncols) {
+            throw new IllegalArgumentException("End column must be in range [startColumn,ncols]");
+        }
+
+        for (int j = startColumn; j < endColumn; j++) {
+            for (int k = colIndex[j]; k < colIndex[j + 1]; k++) {
+                int i = rowIndex[k];
+                action.eval(i, j, x[k]);
+            }
+        }
+    }
+
+    /**
+     * Provides a stream over all of the non-zero elements of a sparse matrix.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public Stream<SparseMatrixEntry> nonzeros() {
```

Comment:
Shall we just call it `stream`? It is more consistent with overall Java Stream design.

Reply:
That would also be misleading since it isn't a stream over the entire matrix. Precedent for a specialized name for a stream can be found in [Files.lines()](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#lines-java.nio.file.Path-)